veracruz-utils/src/policy.rs

```diff
@@ -93,6 +96,133 @@ impl From<x509_parser::error::PEMError> for VeracruzUtilError {
     }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// File operation and capabilities.
+////////////////////////////////////////////////////////////////////////////////
+
+/// List of file operations
+/// TODO: line up wasi-type, if necessary
+#[derive(Clone, Hash, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum VeracruzCapability {
+    Read,
+    Write,
+    Execute,
+}
+
+#[derive(Clone,Hash,PartialEq,Eq,Debug)]
+pub enum VeracruzCapabilityIndex {
+    // Client ID
+    Principal(u64),
+    // Program
+    Program(String),
+}
+
+pub type VeracruzCapabilityTable = HashMap<VeracruzCapabilityIndex,HashMap<String, HashSet<VeracruzCapability>>>;
+
+/// Defines the permission of a file.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct VeracruzFileCapability {
+    /// The file name 
+    file_name : String,
+    /// Read permission
+    read : bool,
+    /// Write permission
+    write : bool,
+    /// Execute permission
+    execute : bool,
+}
+
+impl VeracruzFileCapability {
+    /// Creates a new file permission.
+    #[inline]
+    pub fn new(file_name: String, read: bool, write: bool, execute: bool) -> Self {
+        Self {
+            file_name,
+            read,
+            write,
+            execute,
+        }
+    }
+
+    /// Returns the file_name.
+    #[inline]
+    pub fn file_name(&self) -> &str {
+        self.file_name.as_str()
+    }
+
+    /// If it is allowed to read.
+    #[inline]
+    pub fn read(&self) -> bool {
+        self.read
+    }
+
+    /// If it is allowed to write.
+    #[inline]
+    pub fn write(&self) -> bool {
+        self.write
+    }
+
+    /// If it is allowed to execute.
+    #[inline]
+    pub fn execute(&self) -> bool {
+        self.execute
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Program permission
+////////////////////////////////////////////////////////////////////////////////
+/// Defines a program that can be loaded into execution engine.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct VeracruzProgram {
+    /// The file name 
+    program_file_name : String,
+    /// The program ID
+    id: u32,
```

Comment:
This is half a question, as I've been a bit confused on how data is matched to ids in Durango.

Would it make sense to drop integer ids completely, and instead give each data/program a file name in the filesystem? That way Durango could provide data/streams as a key (file name) value pair.

Reply:
I have not modified Durango yet. If you want to have a play, you can see my current change on `test-template` in `sinaloa-test`.